### DISCUSSION TITLE
Unable to encrypt private key with encryption algo AES-256-CBC using i2d_PKCS8PrivateKey_nid_bio API

### DISCUSSION THREAD

**Discussion by absrivas-in** [openssl/openssl#24440] [Q&A]
Version: OpenSSL 3.0.13 30 

Hi,
When i am trying to decrypt the private key  encrypted with Algo NID_pbe_WithSHA1And3_Key_TripleDES_CBC
using i2d_PKCS8PrivateKey_nid_bio API in fips mode i am getting error. 

In non fips mode it is working fine. 

Hence to allow decryption in fips mode i want to encrypt private key with AES-256-CBC mode using i2d_PKCS8PrivateKey_nid_bio API, but in documentation or openssl sample code i didn't find anything related to my requirement.

Can you please help me on this.

**Comment by absrivas-in** [comment]
Found **i2d_PKCS8PrivateKey_bio** api where i can pass evp aes-256-cbc cipher.
May i know is it right approach for encrypting private key ?

const EVP_CIPHER *cipher = EVP_CIPHER_fetch(libctx, "AES-256-CBC", NULL);
if(i2d_PKCS8PrivateKey_bio(bp, pkeyobj, cipher, (const char*)encpwd, encpwdlen, NULL, NULL) != 1)
{
   printf("i2d_PKCS8PrivateKey_bio Failed");
}
if (cipher != NULL)
      EVP_CIPHER_free(cipher);

**Reply by t8m** [reply]
Yes, this is the right way how to encrypt the private key with AES-256-CBC.

**Reply by absrivas-in** [reply]
Thanks @t8m for confirming.
I have some query on top that and requiring some clarification:

1. Can you please confirm that NID_pbe_WithSHA1And3_Key_TripleDES_CBC is not supported in fips mode.
2.  Why only decryption is failing in fips mode not the encryption with algo NID_pbe_WithSHA1And3_Key_TripleDES_CBC.
3. PKCS12 file decryption in fips mode we are not seeing such error if we pkcs12 is encrypted with
NID_pbe_WithSHA1And3_Key_TripleDES_CBC. Why ?

**Reply by t8m** [reply]
You would have to give concrete commands that you're using with the concrete errors you're seeing. Otherwise I am unable to elaborate this further.

**Reply by absrivas-in** [reply]
Hi @t8m ,
As requested please find following information:

**Code snippet:**
[pvtKeyDecryption.c.zip](https://github.com/user-attachments/files/15742283/pvtKeyDecryption.c.zip)
_Failure is happening in below code snippet._
```
p8inf = PKCS8_decrypt_ex(p8, passphrase, strlen(passphrase), lib_ctx, NULL);
if (p8inf == NULL) {
    fprintf(stderr, "Enable to decrypt Pvt Key !!!\n");
    oret =  -1;
    goto end;
}
```

**Output:**
```
Loading fips provider...
All Provider has been loaded !!!
Enable to decrypt Pvt Key !!!
- Detailed error: unsupported
Private key decryption failed
```

[private_Key.pem.zip](https://github.com/user-attachments/files/15742257/private_Key.pem.zip)
password: keypwd12

Observation:
**openssl version**
_OpenSSL 3.0.13 30 Jan 2024 (Library: OpenSSL 3.0.13 30 Jan 2024)_

`openssl pkcs8 -topk8 -in private_key.pem -passin pass:keypwd12 -outform DER -out private_key.der -v1 PBE-SHA1-DES -passout pass:passwd`
**output:**
`Error encrypting key
80E2169AEF7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:386:Global default library context, Algorithm (PBKDF1 : 0), Properties (<null>)
80E2169AEF7F0000:error:11800067:PKCS12 routines:PKCS12_item_i2d_encrypt_ex:encrypt error:crypto/pkcs12/p12_decr.c:193:
80E2169AEF7F0000:error:11800067:PKCS12 routines:PKCS8_set0_pbe_ex:encrypt error:crypto/pkcs12/p12_p8e.c:80:` 


If i am running same command with following version, i am able to generate der file with pbeWithSHA1AndDES-CBC
**OpenSSL 1.1.1k  FIPS 25 Mar 2021**

**Reply by t8m** [reply]
You need the legacy provider loaded to be able to use `-v1 PBE-SHA1-DES` algorithm. This is an extremely weak algorithm not providing any real protection. I would not recommend using it at all.

**Reply by absrivas-in** [reply]
Hi @t8m , 
PBE-SHA1-DES  transalates to pbeWithSHA1And3-KeyTripleDES-CBC 
1) Why error is not being thrown at the time of encryption in fips mode if it is weak algorithm ?
2) I am seeing different behaviour when trying to load p12 file wallet encrypted with  pbeWithSHA1And3-KeyTripleDES-CBC using following API.
`if (PKCS12_get_safe( asafes, i, (char*)passwd, passwdLen, &bags, &count) != 1)`

By using same algo **pbeWithSHA1And3-KeyTripleDES-CBC**
**PKCS12_get_safe** is able to decrypt p12 file in fips mode. but 
**PKCS8_decrypt_ex** is unable to decrypt the private key in fips mode ?

Is this is a bug in Openssl ?
